scrapy/core/downloader/handlers/http11.py

```diff
@@ -123,83 +127,88 @@ class TunnelingTCP4ClientEndpoint(TCP4ClientEndpoint):
 
     def __init__(
         self,
-        reactor,
-        host,
-        port,
-        proxyConf,
-        contextFactory,
-        timeout=30,
-        bindAddress=None,
+        reactor: ReactorBase,
+        host: str,
+        port: int,
+        proxyConf: Tuple[str, int, Optional[bytes]],
+        contextFactory: IPolicyForHTTPS,
+        timeout: float = 30,
+        bindAddress: Optional[Tuple[str, int]] = None,
     ):
         proxyHost, proxyPort, self._proxyAuthHeader = proxyConf
         super().__init__(reactor, proxyHost, proxyPort, timeout, bindAddress)
-        self._tunnelReadyDeferred = defer.Deferred()
-        self._tunneledHost = host
-        self._tunneledPort = port
-        self._contextFactory = contextFactory
-        self._connectBuffer = bytearray()
+        self._tunnelReadyDeferred: Deferred = Deferred()
+        self._tunneledHost: str = host
+        self._tunneledPort: int = port
+        self._contextFactory: IPolicyForHTTPS = contextFactory
+        self._connectBuffer: bytearray = bytearray()
 
-    def requestTunnel(self, protocol):
+    def requestTunnel(self, protocol: Protocol) -> Protocol:
         """Asks the proxy to open a tunnel."""
+        assert protocol.transport
         tunnelReq = tunnel_request_data(
             self._tunneledHost, self._tunneledPort, self._proxyAuthHeader
         )
         protocol.transport.write(tunnelReq)
         self._protocolDataReceived = protocol.dataReceived
-        protocol.dataReceived = self.processProxyResponse
+        protocol.dataReceived = self.processProxyResponse  # type: ignore[method-assign]
         self._protocol = protocol
         return protocol
 
-    def processProxyResponse(self, rcvd_bytes):
+    def processProxyResponse(self, data: bytes) -> None:
```

Comment:
Technically backward incompatible, but :see_no_evil:

Reply:
I assume it's only called when assigned to `protocol.dataReceived` so there shouldn't be problems even theoretically.